src/components/modules/caret.ts

```diff
@@ -548,6 +546,102 @@ export default class Caret extends Module {
     selection.addRange(newRange);
   }
 
+  /**
+   * Detect next position from current position
+   */
+  private detectNext(): Position | false {
+    const { BlockManager, Tools } = this.Editor;
+
+    const nextInput = BlockManager.currentBlock.nextInput;
+    let nextBlock = BlockManager.nextContentfulBlock;
+
+    if (!nextBlock && !nextInput) {
+      /**
+       * If there is a last block and it is default, do nothing
+       * This code allows to exit from the last non-initial tool:
+       * https://github.com/codex-team/editor.js/issues/1103
+       */
+      if (Tools.isDefault(BlockManager.currentBlock.tool)) {
+        return false;
+      }
+
+      /**
+       * If there is no nextBlock, but currentBlock is not default,
+       * insert new default block at the end and navigate to it
+       */
+      nextBlock = BlockManager.insertAtEnd();
+    }
+
+    return this.detectPosition(true, nextBlock, nextInput);
+  }
+
+  /**
+   * Detect previous position from current position
+   */
+  private detectPrevious(): Position | false {
+    const { BlockManager } = this.Editor;
+
+    const previousInput = BlockManager.currentBlock.previousInput;
+    const previousBlock = BlockManager.previousContentfulBlock;
+
+    if (!previousBlock && !previousInput) {
+      return false;
+    }
+
+    return this.detectPosition(false, previousBlock, previousInput);
+  }
+
+  /**
+   * Detect next or previous position from current position
+   *
+   * @param {boolean} isNext - if true, detect next line. Previous otherwise
+   * @param {Block} block - next or previous Block
+   * @param {HTMLElement | undefined} input - next or previous Block
+   */
+  private detectPosition(isNext: boolean, block: Block, input?: HTMLElement): Position {
+    const currentBoundingClientRect = Selection.get().getRangeAt(0)
+      .getBoundingClientRect();
+
+    const range = new Range();
+    const root = input ?? block.firstInput;
+
+    let offset = isNext ? 0 : root.textContent.length - 1;
+    let prevBoundingClientRect: DOMRect | undefined;
+
+    let position = {
+      block,
+      input,
+      offset: isNext ? root.textContent.length : 0,
+    };
+
+    this.walkTextNodeChars(root, isNext, (textNode, index) => {
```

Comment:
Please add comments for this part of the code

Reply:
Added.